packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -561,15 +561,55 @@ class RenderShaderMask extends RenderProxyBox {
   }
 }
 
+abstract class CustomClipper {
+  Rect getClipRect(Size size);
+  bool shouldRepaint(CustomClipper oldClipper);
+}
+
+class _RenderCustomClip extends RenderProxyBox {
+  _RenderCustomClip({
+    RenderBox child,
+    CustomClipper clipper
+  }) : _clipper = clipper, super(child);
+
+  CustomClipper get clipper => _clipper;
+  CustomClipper _clipper;
+  void set clipper (CustomClipper newClipper) {
+    if (_clipper == newClipper)
+      return;
+    CustomClipper oldClipper = _clipper;
+    _clipper = newClipper;
+    if (newClipper == null) {
+      assert(oldClipper != null);
+      markNeedsPaint();
+    } else if (oldClipper == null
+        || oldClipper.runtimeType != oldClipper.runtimeType
```

Comment:
this is a case where i'm not convinced that the ||-on-the-left pattern is prettier, fwiw